File: plugin-dotnet-server/src/main/java/jetbrains/buildServer/dotnet/DotnetRunnerRunType.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.dotnet;

import jetbrains.buildServer.requirements.Requirement;
import jetbrains.buildServer.requirements.RequirementType;
import jetbrains.buildServer.serverSide.InvalidProperty;
import jetbrains.buildServer.serverSide.PropertiesProcessor;
import jetbrains.buildServer.serverSide.RunType;
import jetbrains.buildServer.serverSide.RunTypeRegistry;
import jetbrains.buildServer.util.StringUtil;
import jetbrains.buildServer.web.openapi.PluginDescriptor;
import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * Dotnet runner definition.
 */
public class DotnetRunnerRunType extends RunType {

    private final PluginDescriptor myPluginDescriptor;

    public DotnetRunnerRunType(@NotNull final PluginDescriptor pluginDescriptor,
                               @NotNull final RunTypeRegistry runTypeRegistry) {
        myPluginDescriptor = pluginDescriptor;
        runTypeRegistry.registerRunType(this);
    }

    @NotNull
    @Override
    public String getType() {
        return DotnetConstants.RUNNER_TYPE;
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return DotnetConstants.RUNNER_DISPLAY_NAME;
    }

    @NotNull
    @Override
    public String getDescription() {
        return DotnetConstants.RUNNER_DESCRIPTION;
    }

    @Override
    public PropertiesProcessor getRunnerPropertiesProcessor() {
        return new PropertiesProcessor() {
            @Override
            public Collection<InvalidProperty> process(Map<String, String> map) {
                return Collections.emptyList();
            }
        };
    }

    @Override
    public String getEditRunnerParamsJspFilePath() {
        return myPluginDescriptor.getPluginResourcesPath("editDotnetParameters.jsp");
    }

    @Override
    public String getViewRunnerParamsJspFilePath() {
        return myPluginDescriptor.getPluginResourcesPath("viewDotnetParameters.jsp");
    }

    @Override
    public Map<String, String> getDefaultRunnerProperties() {
        return new HashMap<String, String>();
    }

    @NotNull
    @Override
    public String describeParameters(@NotNull Map<String, String> parameters) {
        final String paths = parameters.get(DotnetConstants.PARAM_PATHS);
        return String.format("dotnet %s %s",
                parameters.get(DotnetConstants.PARAM_COMMAND),
                StringUtil.isEmpty(paths) ? StringUtil.EMPTY : paths);
    }

    @NotNull
    @Override
    public List<Requirement> getRunnerSpecificRequirements(@NotNull Map<String, String> runParameters) {
        return Collections.singletonList(new Requirement(DotnetConstants.CONFIG_PATH, null, RequirementType.EXISTS));
    }
}

File: plugin-dotnet-server/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetRunnerRunType.kt


package jetbrains.buildServer.dotnet

import jetbrains.buildServer.RequirementsProvider
import jetbrains.buildServer.serverSide.InvalidProperty
import jetbrains.buildServer.serverSide.PropertiesProcessor
import jetbrains.buildServer.serverSide.RunType
import jetbrains.buildServer.serverSide.RunTypeRegistry
import jetbrains.buildServer.util.StringUtil
import jetbrains.buildServer.web.openapi.PluginDescriptor

/**
 * Dotnet runner definition.
 */
class DotnetRunnerRunType(
    runTypeRegistry: RunTypeRegistry,
    private val _pluginDescriptor: PluginDescriptor,
    private val _dotnetRunnerRequirementsProvider: RequirementsProvider,
) : RunType() {

    init {
        _pluginDescriptor.pluginResourcesPath
        runTypeRegistry.registerRunType(this)
    }

    override fun getType(): String {
        return DotnetConstants.RUNNER_TYPE
    }

    override fun getDisplayName(): String {
        return DotnetConstants.RUNNER_DISPLAY_NAME
    }

    override fun getDescription(): String {
        return DotnetConstants.RUNNER_DESCRIPTION
    }

    override fun getRunnerPropertiesProcessor(): PropertiesProcessor {
        return PropertiesProcessor { properties ->
            val command = properties?.get(DotnetConstants.PARAM_COMMAND)
            if (command.isNullOrEmpty()) {
                return@PropertiesProcessor arrayListOf(InvalidProperty(DotnetConstants.PARAM_COMMAND, "Command must be set"))
            }

            val errors = arrayListOf<InvalidProperty>()
            DotnetParametersProvider.commandTypes[command]?.let {
                errors.addAll(it.validateProperties(properties))
            }

            properties[CoverageConstants.PARAM_TYPE]?.let {
                DotnetParametersProvider.coverageTypes[it]?.let {
                    errors.addAll(it.validateProperties(properties))
                }
            }

            errors
        }
    }

    override fun getEditRunnerParamsJspFilePath(): String {
        return _pluginDescriptor.getPluginResourcesPath("editDotnetParameters.jsp")
    }

    override fun getViewRunnerParamsJspFilePath(): String {
        return _pluginDescriptor.getPluginResourcesPath("viewDotnetParameters.jsp")
    }

    override fun getDefaultRunnerProperties() = mapOf(
        CoverageConstants.PARAM_DOTCOVER_MERGE to "true",
        CoverageConstants.PARAM_DOTCOVER_REPORT to "true"
    )

    override fun describeParameters(parameters: Map<String, String>): String {
        val paths = (parameters[DotnetConstants.PARAM_PATHS] ?: "").trim()
        val commandName = parameters[DotnetConstants.PARAM_COMMAND]?.replace('-', ' ')
        val args = parameters[DotnetConstants.PARAM_ARGUMENTS]?.let {
            StringUtil.splitCommandArgumentsAndUnquote(it).take(2).joinToString(" ")
        } ?: ""

        return when {
            commandName == DotnetCommandType.Custom.id -> "$paths\nCommand line parameters: $args"
            !commandName.isNullOrBlank() -> "$commandName $paths"
            else -> args
        }
    }

    override fun getRunnerSpecificRequirements(runParameters: Map<String, String>) =
        _dotnetRunnerRequirementsProvider.getRequirements(runParameters).toList()

    override fun getTags(): MutableSet<String> {
        return mutableSetOf(".NET", "MSBuild", "VSTest", "VS", "Visual Studio", "NuGet", "devenv")
    }

    override fun getIconUrl(): String {
        return _pluginDescriptor.getPluginResourcesPath("dotnet-runner.svg");
    }
}